src/database.rs

```diff
@@ -499,183 +519,205 @@ impl Database {
-    pub fn scope_result(&mut self) -> &Arc<ScopeResult> {
+    pub fn scope_pass(&mut self) -> &Arc<ScopeResult> {
         if self.scopes.is_none() {
-            self.name_result();
+            self.name_pass();
             time(&self.options.clone(), "scopeck", || {
-                if self.prev_scopes.is_none() {
-                    self.prev_scopes = Some(Arc::new(ScopeResult::default()));
-                }
-
-                let parse = self.parse_result().clone();
-                let name = self.name_result().clone();
-                {
-                    let ns = Arc::make_mut(self.prev_scopes.as_mut().unwrap());
-                    scopeck::scope_check(ns, &parse, &name);
-                }
-                self.scopes = self.prev_scopes.clone();
+                let mut sc = self.prev_scopes.take().unwrap_or_default();
+                let parse = self.parse_result();
+                let name = self.name_result();
+                scopeck::scope_check(Arc::make_mut(&mut sc), parse, name);
+                self.prev_scopes = Some(sc.clone());
+                self.scopes = Some(sc);
             });
         }
+        self.scope_result()
+    }
 
+    /// Returns the frames for this database, i.e. the actual logical system.
+    /// Panics if [`Database::scope_pass`] was not previously called.
+    ///
+    /// All logical properties of the database (as opposed to surface syntactic
+    /// properties) can be obtained from this object.
+    #[must_use]
+    pub fn scope_result(&self) -> &Arc<ScopeResult> {
         self.scopes.as_ref().unwrap()
     }
 
     /// Calculates and returns verification information for the database.
     ///
     /// This is an optimized verifier which returns no useful information other
     /// than error diagnostics.  It does not save any parsed proof data.
-    pub fn verify_result(&mut self) -> &Arc<VerifyResult> {
+    pub fn verify_pass(&mut self) -> &Arc<VerifyResult> {
         if self.verify.is_none() {
-            self.name_result();
-            self.scope_result();
+            self.name_pass();
+            self.scope_pass();
             time(&self.options.clone(), "verify", || {
-                if self.prev_verify.is_none() {
-                    self.prev_verify = Some(Arc::new(VerifyResult::default()));
-                }
-
-                let parse = self.parse_result().clone();
-                let scope = self.scope_result().clone();
-                let name = self.name_result().clone();
-                {
-                    let ver = Arc::make_mut(self.prev_verify.as_mut().unwrap());
-                    verify::verify(ver, &parse, &name, &scope);
-                }
-                self.verify = self.prev_verify.clone();
+                let mut ver = self.prev_verify.take().unwrap_or_default();
+                let parse = self.parse_result();
+                let scope = self.scope_result();
+                let name = self.name_result();
```

Comment:
The problem with the values returned by `name_pass` and the other similar functions is that the returned value has a lifetime tied to an exclusive borrow of `self`, so it can't be used if you want to modify `self` afterward (in this case for the `self.prev_verify.take()` call). Calling `self.name_result` doesn't have that issue since it uses a shared borrow of `self`.